tachyon/math/polynomials/univariate/radix2_dit_parallel.h

```diff
@@ -0,0 +1,224 @@
+    ReverseSliceIndexBits(twiddles);
+    ParDitLayerRev(mat, mid, twiddles);
+    ReverseMatrixIndexBits(mat);
+  }
+
+  static void CosetLDEBatch(RowMajorMatrix<F>& mat, size_t added_bits,
+                            const F& shift) {
+    size_t r = static_cast<size_t>(mat.rows());
+    size_t log_r = base::bits::StrictLog2(r);
+    size_t mid = log_r / 2;
+    std::optional<F> r_inv = F(r).Inverse();
+    CHECK(r_inv);
+
+    F root = TwoAdicGenerator(log_r);
+    std::optional<F> root_inv = root.Inverse();
+    CHECK(root_inv);
+    std::vector<F> twiddles_inv =
+        F::GetSuccessivePowers(r == 0 ? 1 : r / 2, *root_inv);
+
+    // The first half looks like a normal DIT.
+    ReverseMatrixIndexBits(mat);
+    ParDitLayer(mat, mid, twiddles_inv);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    ReverseSliceIndexBits(twiddles_inv);
+    ParDitLayerRev(mat, mid, twiddles_inv);
+    // We skip the final bit-reversal, since the next FFT expects bit-reversed
+    // input.
+
+    // Rescale coefficients in two ways:
+    // - divide by number of rows (since we're doing an inverse DFT)
+    // - multiply by powers of the coset shift (see default coset LDE impl for
+    // an explanation)
+    const std::vector<F> weights = F::GetSuccessivePowers(r, shift, *r_inv);
+    OPENMP_PARALLEL_FOR(size_t row = 0; row < weights.size(); ++row) {
+      // reverse_bits because mat is encoded in bit-reversed order
+      mat.row(base::bits::BitRev(row) >>
+              (sizeof(size_t) * 8 - base::bits::CountTrailingZeroBits(r))) *=
+          weights[row];
+    }
+    BitReversedZeroPad(mat, added_bits);
+
+    r = static_cast<size_t>(mat.rows());
+    log_r = base::bits::StrictLog2(r);
+    mid = log_r / 2;
+
+    root = TwoAdicGenerator(log_r);
+    std::vector<F> twiddles = F::GetSuccessivePowers(r / 2, root);
+
+    // The first half looks like a normal DIT.
+    ParDitLayer(mat, mid, twiddles);
+
+    // For the second half, we flip the DIT, working in bit-reversed order.
+    ReverseMatrixIndexBits(mat);
+    ReverseSliceIndexBits(twiddles);
+    ParDitLayerRev(mat, mid, twiddles);
+    ReverseMatrixIndexBits(mat);
+  }
+
+ private:
+  // This can be used as the first half of a parallelized butterfly network.
+  static void ParDitLayer(RowMajorMatrix<F>& mat, size_t mid,
+                          const std::vector<F>& twiddles) {
+    size_t rows = static_cast<size_t>(mat.rows());
+    size_t cols = static_cast<size_t>(mat.cols());
+    size_t log_r = base::bits::StrictLog2(static_cast<size_t>(mat.rows()));
+    size_t chunk_rows = 1 << mid;
+
+    // max block size: 2^mid
+    OPENMP_PARALLEL_FOR(size_t block_start = 0; block_start < rows;
+                        block_start += chunk_rows) {
+      size_t cur_chunk_rows = std::min(chunk_rows, rows - block_start);
+      RowMajorMatrix<F> submat =
+          mat.block(block_start, 0, cur_chunk_rows, cols);
```

Comment:
This will performs a copy. so please use the type of the return type of `mat.block()` explicitly